kernel/src/transaction.rs

```diff
@@ -1,120 +1,249 @@
 use std::sync::Arc;
 
-use crate::actions::visitors::TransactionVisitor;
-use crate::actions::{get_log_schema, TRANSACTION_NAME};
+use crate::actions::get_log_schema;
+use crate::schema::{Schema, SchemaRef, StructType};
 use crate::snapshot::Snapshot;
-use crate::Engine;
-use crate::{actions::Transaction, DeltaResult};
+use crate::{DataType, Expression};
+use crate::{DeltaResult, Engine, EngineData};
 
-pub use crate::actions::visitors::TransactionMap;
-pub struct TransactionScanner {
-    snapshot: Arc<Snapshot>,
+const KERNEL_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+/// A transaction represents an in-progress write to a table.
+pub struct Transaction {
+    read_snapshot: Arc<Snapshot>,
+    commit_info: Option<EngineCommitInfo>,
+}
+
+// Since the engine can include any commit info it likes, we unify the data/schema pair as a single
+// struct with Arc semantics.
+#[derive(Clone)]
+struct EngineCommitInfo {
+    data: Arc<dyn EngineData>,
+    schema: SchemaRef,
 }
 
-impl TransactionScanner {
-    pub fn new(snapshot: Arc<Snapshot>) -> Self {
-        TransactionScanner { snapshot }
+impl std::fmt::Debug for Transaction {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&format!(
+            "Transaction {{ read_snapshot version: {}, commit_info: {} }}",
```

Comment:
this specific implementation doesn't but it is all handled in the `Debug` trait via Formatter's [`alternate()`](https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.alternate). if we wanted to handle pretty printing it would be something like
```
fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    if f.alternate() {
        // Pretty-print formatting when using {:#?}
    } else {
        // Normal formatting
        f.write_str(...);
    }
}
```

Reply:
TIL, thanks!

Reply:
(resolving this just to clean up the PR :) )